### PR TITLE
Add search for DomainData objects

### PR DESCRIPTION
**Pull request by ericholscher**
This adds search indexing for DomainData objects, along with cleaning up our existing search. The major changes are:

* Adds Code API types to the site and project search. The UX of these results is similar to how Sphinx displays them. 
* Removes the `elastic_project_search` view, and reuses the same view & templates as the site search. This also means we now have faceted search for dashboard project search 🎉 
* Adds an experiment `AllSearch` which searches across all our indexes to return results. I'm hoping to eventually move to this as a default search type, so we can display all the types. This also allows us to _facet_ by document type, which is pretty neat. 

### v2

A v2 of this will integrate the Code API & Page search results into one stream, which is what the `AllSearch` is currently doing -- it's implemented and working on the site search, but the UX is not ready to show to users (you can hit it by manually changing your `type` to `all` in the URL). 

The v2 for integrating this into the docsearch API and indoc search will likely require #5235 first, so that we can use multiple serializers for the data, so it's a larger refactor. 

Unifying the modeling between ImportedFile and SphinxDomain is also a goal. We likely need another model as well for Sphinx Domain Type's (eg. `std:doc` or `py:function`, that keep track of metadata about them (eg. `display_name`). That will help make this data feel much more nicely factored. 

Depends on https://github.com/rtfd/readthedocs-sphinx-ext/pull/62

Closes https://github.com/rtfd/readthedocs.org/issues/4289

### Site domain search

screenshot 2019-03-01 11 30 21

### Project search

screenshot 2019-03-01 11 31 24

### CODE REVIEW

**Comment by stsewd** [comment]
Now we can change the base branch to master here I guess

File: readthedocs/projects/tasks.py [hunk 1]
```diff
@@ -1351,7 +1384,7 @@ def _manage_imported_files(version, path, commit):
                 created_html_files.append(obj)
 
     # Send bulk_post_create signal for bulk indexing to Elasticsearch
-    bulk_post_create.send(sender=HTMLFile, instance_list=created_html_files)
+    bulk_post_create.send(sender=PageDocument, instance_list=created_html_files)
```
Comment (safwanrahman) [inline readthedocs/projects/tasks.py]: Whats the reason to change the `sender` from model instanc to document instance? the `instance_list` is `HTMLFile` object list, but you are sending `PageDocument` as sender.

File: readthedocs/projects/tasks.py [hunk 2]
```diff
@@ -1288,22 +1303,40 @@ def warn(self, msg):
...
             if obj.commit != commit:
                 obj.commit = commit
                 obj.save()
-    SphinxDomain.objects.filter(project=version.project,
-                                version=version
-                                ).exclude(commit=commit).delete()
+            created_sphinx_domains.append(obj)
+
+    # Send bulk_post_create signal for bulk indexing to Elasticsearch
+    bulk_post_create.send(sender=SphinxDomainDocument, instance_list=created_sphinx_domains)
```
Comment (safwanrahman) [inline readthedocs/projects/tasks.py]: I think you should send the signal with the `SphinxDomain` as sender.

File: readthedocs/search/faceted_search.py [hunk 3]
```diff
@@ -20,19 +43,37 @@ def __init__(self, user, **kwargs):
             The `self.user` attribute isn't currently used on the .org,
             but is used on the .com
         """
-        self.user = user
-        self.filter_by_user = kwargs.pop('filter_by_user', None)
-        super().__init__(**kwargs)
+        s = super().search()
+        s = s.source(exclude=['content', 'headers'])
+        # Return 25 results
+        return s[:25]
```
Comment (safwanrahman) [inline readthedocs/search/faceted_search.py]: I think this limiting result should be done in the consumer end, like in view end. Doing this in here will limit the usage of the `search` api.

File: readthedocs/search/tasks.py [hunk 4]
```diff
@@ -106,7 +106,8 @@ def index_missing_objects(app_label, model_name, document_class, index_generatio
     """
     model = apps.get_model(app_label, model_name)
     document = _get_document(model=model, document_class=document_class)
-    queryset = document().get_queryset().exclude(modified_date__lte=index_generation_time)
+    query_string = '{}__lte'.format(document.modified_model_field)
+    queryset = document().get_queryset().exclude(**{query_string: index_generation_time})
```
Comment (safwanrahman) [inline readthedocs/search/tasks.py]: Good thing done! I have done this thing in kitsune project.

File: readthedocs/projects/tasks.py [hunk 5]
```diff
@@ -1288,22 +1303,40 @@ def warn(self, msg):
...
             display_name = einfo[3]
-            obj, _ = SphinxDomain.objects.get_or_create(
+            obj, created = SphinxDomain.objects.get_or_create(
                 project=version.project,
                 version=version,
                 domain=domain,
                 name=name,
                 display_name=display_name,
                 type=_type,
+                type_display=types.get(domain + ':' + _type, ''),
```
Comment (stsewd) [inline readthedocs/projects/tasks.py]: We can use f-stirngs here :blush:

File: readthedocs/projects/tasks.py [hunk 6]
```diff
@@ -1288,22 +1303,40 @@ def warn(self, msg):
...
+                doc_display=titles.get(doc_name, ''),
                 anchor=anchor,
             )
             if obj.commit != commit:
                 obj.commit = commit
                 obj.save()
-    SphinxDomain.objects.filter(project=version.project,
-                                version=version
-                                ).exclude(commit=commit).delete()
+            created_sphinx_domains.append(obj)
```
Comment (stsewd) [inline readthedocs/projects/tasks.py]: Don't we need to check `if created` before adding `obj`?

File: readthedocs/settings/base.py [hunk 7]
```diff
@@ -327,8 +328,8 @@ def USE_PROMOS(self):  # noqa
 
     # CORS
     CORS_ORIGIN_REGEX_WHITELIST = (
-        r'^http://(.+)\.readthedocs\.io$',
-        r'^https://(.+)\.readthedocs\.io$',
+        '^http://(.+)\.readthedocs\.io$',
+        '^https://(.+)\.readthedocs\.io$'
```
Comment (stsewd) [inline readthedocs/settings/base.py]: Why are we removing the `raw string` prefix?

**Comment by ericholscher** [comment]
Going to merge this so it will go out in the next deploy 🎉

**Pull request merged**